[3D Map] Cross section nudging

ViperMiniQ opened the pull request:
This PR implements cross section nudging in the similar way the elevation profile nudging works.

The user is able to set the tolerance (the width of the cross section) and nudging it left or right offsets the rubberband polygon and the cross section by that amount.

[Screencast_20251215_102832.webm](https://github.com/user-attachments/assets/d542cb23-dabf-41a4-87e0-7ed9251a34dc)

We are keeping the current functionality as it is, the user is still able to dynamically set the width of the polygon by drawing the third point on the canvas. 

The user is able to nudge the canvas rubberband polygon by pressing the nudge left/right buttons in the menu, or by using keyboard shortcuts (customizable from the shortcuts manager) when the menu is not being shown.

ViperMiniQ committed:
add gui actions

DelazJ commented:
@ViperMiniQ So the last six options are related to cross section? Is that not too much for that menu? Moreover can you please screen cast how users could e.g. nudge three times the extent to the left? I mean, from memory, that drop-down menu closes after you interact with it, one would then need to reopen it between two nudge clicks?

ViperMiniQ committed:
fix locking behavior, clear geometry when cross section disabled

ViperMiniQ committed:
nudge camera when nudging curve, same offset

wonder-sk commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -62,6 +63,9 @@
 #include "moc_qgs3dmapcanvaswidget.cpp"
 
+const QgsSettingsEntryDouble *Qgs3DMapCanvasWidget::settingClippingTolerance = new QgsSettingsEntryDouble( QStringLiteral( "tolerance" ), QgsSettingsTree::sTree3DMap, 100, QStringLiteral( "Tolerance distance for 3D Map cross section" ), Qgis::SettingsOptions(), 0 );
+const QgsSettingsEntryBool *Qgs3DMapCanvasWidget::settingDynamicClipping = new QgsSettingsEntryBool( QStringLiteral( "enable-dynamic-clipping" ), QgsSettingsTree::sTree3DMap, true, QStringLiteral( "Whether dynamic clipping is set for 3D cross section" ) );
```
I would suggest to rename "dynamic clipping" as it is unclear what that means... for the label, how about "Cross section tolerance lock - whether the tolerance is set by mouse every time a new cross section is picked"

wonder-sk commented on src/app/3d/qgsmaptoolclippingplanes.cpp:
```diff
@@ -127,7 +127,34 @@ void QgsMapToolClippingPlanes::canvasReleaseEvent( QgsMapMouseEvent *e )
+        QgsGeometry geom( new QgsLineString( points3DMap ) );
+        geom.transform( *mCt, Qgis::TransformDirection::Reverse );
+
+        emit finishedSuccessfully( geom, 0 );
```
hmm... we are transforming from 3D map's CRS back to 2D map canvas CRS - but then in the other code in 3D, we are using that geometry unchanged?

I am thinking about the following approach:
- have a new small class that would encapsulate the whole cross section setup: i.e. the two endpoints and width - and it would be always in the 3D map CRS (at some point later we may want to add also Z range to define top and bottom limit). The class could have a function to e.g. generate QgsGeometry for the 2D canvas rubber band
- the map tool could call Qgs3DMapCanvas::setCrossSection( const QgsCrossSection & crossSection ), and the finishedSuccessfully() signal could be without and arguments, and it would only take care of raising the 3D widget back

ViperMiniQ committed:
add tolerance

dynamic polygon selection is still available
tolerance controls the polygon width
nudging offsets rubber band polygon by tolerance value
new settings tree is added for the tolerance setting

ViperMiniQ committed:
create separate button/menu for the cross section

expand tolerance widget logic, add toolbutton to toggle between using set tolerance or user canvas drawn polygon

ViperMiniQ committed:
add customizable shortcuts for nudging

ViperMiniQ committed:
set default shortcuts

ViperMiniQ committed:
set default shortcuts 2

ViperMiniQ committed:
add QgsCrossSection

instead of passing start point, end point and tolerance separately, we encapsulate them all in a qgscrosssection class which we use to set the cross section definition for the 3D map

additionaly, moveCameraPositionBy has been renamed to moveCenterPoint and it has been moved from private to public visibility in qgscameracontroller

ViperMiniQ committed:
include cross section sip files

ViperMiniQ committed:
generate sip files

ViperMiniQ committed:
fix leftover renames of dynamic clipping

ViperMiniQ committed:
update qgscrosssection docs

ViperMiniQ committed:
document method

ViperMiniQ committed:
silence clang tidy, add since tag to docstring

ViperMiniQ committed:
fix docs, silence clang tidy

wonder-sk commented on src/3d/qgs3dmapcanvas.h:
```diff
@@ -147,12 +145,26 @@ class _3D_EXPORT Qgs3DMapCanvas : public QWindow
      */
-    void setRootEntity( Qt3DCore::QEntity *root );
+    void
+      setRootEntity( Qt3DCore::QEntity *root );
```
this looks like unintended change? 🙂

wonder-sk reviewed (commented):
Nearly there!

ViperMiniQ committed:
review fixes

- cross section side view has been moved to cameracontroller
- cross section class has been expanded with isValid method
- disableCrossSection and enableCrossSection have been unified under setCrossSection
- missing menu items has been fixed
- remove \since tags from new class

ViperMiniQ committed:
fix docs, remove sip_skip

ViperMiniQ committed:
remove mention of removed method

ViperMiniQ commented:
Hi @ptitjano,
thanks for the report! See fix in #64654
